src/devices/Is31fl3731/Is31Fl3731.cs

```diff
@@ -0,0 +1,278 @@
+        private const byte PWM_REGISTER = 0x24;
+        private const byte PWM_REGISTER_END = 0xB3;
+        private const byte PWM_REGISTER_LENGTH = (PWM_REGISTER_END - PWM_REGISTER) + 1;
+
+        // Function register
+        // Writes data to page nine
+        // table 3 in datasheet
+        private const byte CONFIGURATION_REGISTER = 0x0;
+        private const byte DISPLAY_REGISTER = 0x5;
+        private const byte SHUTDOWN = 0x0A;
+
+        // Values
+        private readonly byte[] _disable_all_leds_data = new byte[LED_REGISTER_LENGTH];
+        private readonly byte[] _enable_all_leds_data = new byte[LED_REGISTER_LENGTH];
+
+        private I2cDevice _i2cDevice;
+
+        /// <summary>
+        /// Initialize IS31FL3731 device
+        /// </summary>
+        /// <param name="i2cDevice">The <see cref="System.Device.I2c.I2cDevice"/> to create with.</param>
+        public Is31fl3731(I2cDevice i2cDevice)
+        {
+            _i2cDevice = i2cDevice;
+            _enable_all_leds_data.AsSpan().Fill(0xff);
+        }
+
+        /// <summary>
+        /// Initialize IS31FL3731 device
+        /// </summary>
+        /// <param name="i2cDevice">The <see cref="System.Device.I2c.I2cDevice"/> to create with.</param>
+        /// <param name="width">The width of the LED matrix.</param>
+        /// <param name="height">The height of the LED matrix.</param>
+        public Is31fl3731(I2cDevice i2cDevice, int width, int height)
+        : this(i2cDevice)
+        {
+            Width = width;
+            Height = height;
+        }
+
+        /// <summary>
+        /// Default I2C address for device.
+        /// </summary>
+        public static readonly int DefaultI2cAddress;
+
+        /// <summary>
+        /// Width of LED matrix (x axis).
+        /// </summary>
+        public readonly int Width = 16;
+
+        /// <summary>
+        /// Height of LED matrix (y axis).
+        /// </summary>
+        public readonly int Height = 9;
+
+        /// <summary>
+        /// Initialize LED driver.
+        /// </summary>
+        public void Initialize()
+        {
+            // Reset device
+            Reset();
+
+            // set display mode (bits d4:d3)
+            // 00 = picture mode
+            // 01 = auto frame
+            // 1x audio frame play mode
+            Write(FUNCTION_REGISTER, CONFIGURATION_REGISTER, 0);
+            // set data page
+            _i2cDevice.Write(new byte[] { COMMAND_REGISTER, 0 });
+        }
+
+        /// <summary>
+        /// Indexer for updating matrix, with PWM register.
+        /// </summary>
+        public int this[int x, int y]
+        {
+            get => ReadLedPwm(x, y);
+            set => WriteLedPwm(x, y, value);
+        }
+
+        /// <summary>
+        /// Set value for LED.
+        /// </summary>
+        public void WritePixel(int x, int y, int brightness, bool enable, bool blink)
+        {
+            WriteLedBlink(x, y, blink);
```

Comment:
Good question. It's basically impossible for this code to know what values are in/out of range since there are multiple addressing schemes that are used higher-up. Certainly, the code could check for values are obviously out-of-range, like 1000, but I don't think that would have much value since values that are in-range for one matrix are out-of-range for another. This check should be done a layer up.